kube-controllers/cmd/kube-controllers/main.go

```diff
@@ -311,6 +300,19 @@ func runHealthChecks(ctx context.Context, s *status.Status, k8sClientset *kubern
 			s.SetReady("KubeAPIServer", true, "")
 		}
 
+		// If we encountered errors, retry again with a longer timeout.
+		if !s.GetReadiness() {
+			timeout = 2 * timeout
+			if timeout > maxTimeout {
+				timeout = maxTimeout
+			}
+			log.Info("Health check is not ready, retrying in 2 seconds with new timeout: %s", timeout)
+			time.Sleep(2 * time.Second)
+			continue
+		}
+
+		// Success! Reset the timeout.
+		timeout = defaultTimeout
 		time.Sleep(10 * time.Second)
```

Comment:
Is this 10s sleep still needed alongside the exponential backoff? Seems redundant.

Reply:
There are two time durations at play here:

- The timeout we use for our HTTP requests
- The period at which we perform the HTTP requests.

The exponential backoff is only for the first of those two. This 10s sleep is the time we wait in between successful checks (i.e., after 10s we double check that things are still OK).

The exponential backoff added in this PR is only used if we fail a health check, so that on the next iteration we give our HTTP requests a bit longer to execute.

Reply:
Ah, the timeout-ception had me confused, thanks for clarification.